### PR TITLE
=str Add dedicated stream timeout exceptions for timeout related oper…

### PR DESCRIPTION
**Pull request by He-Pin**
Add dedicated stream timeout exceptions for timeout related operators。

refs: https://github.com/akka/akka/issues/31455
refs:https://github.com/akka/akka/issues/24951

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream
+
+import scala.concurrent.TimeoutException
+import scala.util.control.NoStackTrace
+
+sealed class StreamTimeoutException(msg: String) extends TimeoutException(msg) with NoStackTrace
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala:10]: Becomes public from java, so also add:

```suggestion
/**
 * Base class for exceptions specific to Akka Streams
 * 
 * Not for user extension
 */
@DoNotInherit
sealed class StreamTimeoutException(msg: String) extends TimeoutException(msg) with NoStackTrace
```

**Review by johanandren — COMMENTED** [review]
One more thing, then I think it looks good

File: akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream
+
+import scala.concurrent.TimeoutException
+import scala.util.control.NoStackTrace
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala:8]:
```suggestion
import scala.util.control.NoStackTrace
import akka.annotation.DoNotInherit
```

File: akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * Not for user extension
+ */
+@DoNotInherit
+sealed class StreamTimeoutException(msg: String) extends TimeoutException(msg) with NoStackTrace
+
+final class InitialTimeoutException(msg: String) extends StreamTimeoutException(msg)
+
+final class CompletionTimeoutException(msg: String) extends StreamTimeoutException(msg)
+
+final class IdleTimeoutException(msg: String) extends StreamTimeoutException(msg)
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala:23]: There is already an `TcpIdleTimeoutException` , should this be `StreamIdleTimeoutException`?

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala:23]: Makes sense to be explicit 👍

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala:23]: How about the others?

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/Timers.scala [hunk 4]
```diff
@@ -59,12 +59,12 @@ import akka.stream.stage._
 
         final override protected def onTimer(key: Any): Unit =
           if (!initialHasPassed)
-            failStage(new TimeoutException(s"The first element has not yet passed through in $timeout."))
+            failStage(new InitialTimeoutException(s"The first element has not yet passed through in $timeout."))
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/Timers.scala:62]: `toString` of `FiniteDuration` isn't always very readable. Use `timeout.toCoarsest`

File: akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * Not for user extension
+ */
+@DoNotInherit
+sealed class StreamTimeoutException(msg: String) extends TimeoutException(msg) with NoStackTrace
+
+final class InitialTimeoutException(msg: String) extends StreamTimeoutException(msg)
+
+final class CompletionTimeoutException(msg: String) extends StreamTimeoutException(msg)
+
+final class StreamIdleTimeoutException(msg: String) extends StreamTimeoutException(msg)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala:23]: Why is this prefixed with Stream? Others are not. I think we can skip Stream here.

**Review by patriknw — COMMENTED** [review]
looking good, but something small...

File: akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala (hunk 5 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala:23]: There is already an TcpIdleTimeoutException , should this be StreamIdleTimeoutException?

File: akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala (hunk 3 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala]: @patriknw Should we revert it back to `IdleTimeoutException`?

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala]: I see, let's keep it as is then. `StreamIdleTimeoutException`

File: akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala (hunk 5 above)
Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala:23]: this was already discussed https://github.com/akka/akka/pull/31640#pullrequestreview-1130828577

File: akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala (hunk 3 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/StreamTimeoutException.scala]: Okay, let me update the PR to the final state.

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**